Moving to C Bindings instead of deprecated C++ bindings for MPI

henryiii opened the pull request:
This was applied from a `.patch` from the PR GooFit/Minuit2#1 from @gudlaugu. All credit/blame should be directed to @gudlaugu, I'm just the middleman. This patch is useful because apparently MPI 3 no longer even includes support for the C++ bindings.

Once #1680 is merged, this will be much easier to test.

For future reference, the ugly use of environment variables instead of CMake option for MPI options in the main ROOT::Minuit2 CMakeLists.txt needs to be fixed, and also could be moved to FindMPI instead of the current hardcoded method.

amadio commented:
Could you please assign authorship to the right person? Just a
```
$ git commit --amend --author="<author info>"
```
should do. Thanks.

henryiii commented:
Fixed.

amadio commented:
Thanks!

amadio commented:
FYI, I see many warnings like this in your branch:
```
[ 96%] Building CXX object math/minuit2/CMakeFiles/Minuit2.dir/src/Numerical2PGradientCalculator.cxx.o
In file included from /home/amadio/src/root/math/minuit2/src/HessianGradientCalculator.cxx:27:0:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h: In member function ‘unsigned int ROOT::Minuit2::MPIProcess::StartElementIndex() const’:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h:66:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       { return ((fRank<NumElements4JobOut()) ? (fRank*NumElements4Job(fRank)) :
                  ~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/amadio/src/root/math/minuit2/src/MnHesse.cxx:31:0:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h: In member function ‘unsigned int ROOT::Minuit2::MPIProcess::StartElementIndex() const’:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h:66:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       { return ((fRank<NumElements4JobOut()) ? (fRank*NumElements4Job(fRank)) :
                  ~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/amadio/src/root/math/minuit2/src/Numerical2PGradientCalculator.cxx:34:0:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h: In member function ‘unsigned int ROOT::Minuit2::MPIProcess::StartElementIndex() const’:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h:66:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       { return ((fRank<NumElements4JobOut()) ? (fRank*NumElements4Job(fRank)) :
                  ~~~~~^~~~~~~~~~~~~~~~~~~~~
In file included from /home/amadio/build/gcc7.3/math/minuit2/G__Minuit2.cxx:84:0:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h: In member function ‘unsigned int ROOT::Minuit2::MPIProcess::StartElementIndex() const’:
/home/amadio/build/gcc7.3/include/Minuit2/MPIProcess.h:66:23: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       { return ((fRank<NumElements4JobOut()) ? (fRank*NumElements4Job(fRank)) :
                  ~~~~~^~~~~~~~~~~~~~~~~~~~~
```

amadio commented:
Also, `ctest -R minuit2` gives me
```
82% tests passed, 3 tests failed out of 17

Label Time Summary:
tutorial    =   5.25 sec*proc (3 tests)

Total Test time (real) =  19.52 sec

The following tests FAILED:
	 80 - minuit2-Quad1FMain (Failed)
	 90 - minuit2-ReneTest (Failed)
	 93 - minuit2-testMinimizer (Failed)
Errors while running CTest
```
However, these may not be due to your changes. The first of the 3 tests above also fails on the master branch for me.

henryiii commented:
It might be better to use a temporary to get the rank, then assign it to unsigned int rank.

henryiii commented:
Fixed (in theory) the warnings, and moved to using newer style build system (will be integrated with #1680, since there's now some duplication). Uses `MINUIT2_MPI` and `MINUIT2_OMP` instead of environment variables.

amadio commented on math/minuit2/CMakeLists.txt:
```diff
@@ -4,23 +4,62 @@
+    # set(CMAKE_CXX_COMPILER mpic++)
+    # set(CMAKE_C_COMPILER mpic++)
+    # set(CMAKE_CXX_LINK_EXECUTABLE mpic++)
+endif()
```
I've already made very similar changes to this locally, so that we can use this for both ROOT and the standalone Minuit2. Once I merge, I will test and merge the previous commit from this PR. Please wait a bit until I try to fix the broken test, then we can move forward with this.

henryiii replied:
Okay. I have a general idea of how to integrate the two CMake's once this is done.

amadio commented:
@henryiii I was about to merge this locally, but when I tested it, I realized that the changes introduced here (porting to use C bindings for MPI, but not only) actually breaks more tests. I think we will have rework the changes and separate the move to C bindings from other functional changes that are currently in the same commit.

amadio commented:
Another thing, merge commits are not allowed. Please rebase and force-push instead.

henryiii commented:
@amadio Would those be the changes in MnPrint, or more than that? That's all I see that looks like a functional change from a quick check.

henryiii commented:
@amadio, what tests are you running for MPI?

amadio commented:
@henryiii If you configure ROOT with `-Dtesting=ON -Droottest=ON`, then you can run the tests with `ctest -R minuit2`, but it may also be important to run `ctest -R roostats` (or all tests, really, since some mathcore tests also use Minuit2). However, to see the failures, just the Minuit2 tests are enough.

amadio commented:
@phsft-bot build

amadio commented:
Merged manually after fixing some trailing whitespace locally.

amadio commented:
Actually, the PR I merged was #1809, not this one. They were similar enough that I got confused.

amadio commented:
In any case, the commit which moves to C bindings has problems and cannot be merged before the bugs are understood.

henryiii commented:
That's good, the other one was the one that needed.

amadio commented:
This needs to be tested locally, as I do not think we have MPI installed in the externals.

henryiii commented:
That's fine; I'd recommend reopening so that my rebased commits will show up (I'm guessing that's why they are not listed?), and then starting fixes based on that (though understanding the old MPI master might help as well).

henryiii commented:
Didn't know that! Will do. Thanks.
